apps/system/js/nfc_manager.js

```diff
@@ -358,21 +277,31 @@
           if (activeApp.isTransitioning() || activeApp.isSheetTransitioning()) {
             return;
           }
-          // Top visible application's manifest Url is registered;
-          // Start Shrink / P2P UI and wait for user to accept P2P event
-          window.dispatchEvent(new CustomEvent('shrinking-start'));
 
-          // Setup listener for user response on P2P UI now
-          window.addEventListener('shrinking-sent', this);
+          this._initP2PUI();
         } else {
-          // Clean up P2P UI events
-          this._logVisibly('CheckP2PRegistration failed');
-          window.removeEventListener('shrinking-sent', this);
-          window.dispatchEvent(new CustomEvent('shrinking-stop'));
+          this.debug('CheckP2PRegistration failed');
+          this._cleanP2PUI();
         }
       });
     },
 
+    _cleanP2PUI: function() {
+      window.removeEventListener('shrinking-sent', this._handleShrinkingSent);
+      window.dispatchEvent(new CustomEvent('shrinking-stop'));
+    },
+
+    _initP2PUI: function() {
+      window.dispatchEvent(new CustomEvent('shrinking-start'));
+
+      this._handleShrinkingSent = () => {
+        this._cleanP2PUI();
+        this._dispatchP2PUserResponse();
+      };
+
+      window.addEventListener('shrinking-sent', this._handleShrinkingSent);
```

Comment:
I think we should keep this the way it is for now. This is part of the (too-complicated) sharing flow. I think that If we would listen all the time for ```shrinking-sent``` this would mean that ```ShrinkingUI``` would be controlling the nfc sharing, and could trigger not necessary requests to Gecko (by mistake, or misuse).
Since we have ```onpeerfound``` available I think we could start to investigate using it in gaia apps, and let the apps control ```ShrinkingUI```. As result we could remove P2P related code from NfcManage. I'll file a followup bug for that.